src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/EnumConverter.cs

```diff
@@ -16,32 +18,47 @@ internal class EnumConverter<T> : JsonConverter<T>
         // Odd type codes are conveniently signed types (for enum backing types).
         private static readonly string? s_negativeSign = ((int)s_enumTypeCode % 2) == 0 ? null : NumberFormatInfo.CurrentInfo.NegativeSign;
 
+        private const string ValueSeparator = ", ";
+
         private readonly EnumConverterOptions _converterOptions;
-        private readonly JsonNamingPolicy _namingPolicy;
-        private readonly ConcurrentDictionary<string, string>? _nameCache;
+        private readonly JsonNamingPolicy? _namingPolicy;
+        private readonly ConcurrentDictionary<ulong, JsonEncodedText> _nameCache;
 
         public override bool CanConvert(Type type)
         {
             return type.IsEnum;
         }
 
-        public EnumConverter(EnumConverterOptions options)
-            : this(options, namingPolicy: null)
+        public EnumConverter(EnumConverterOptions converterOptions, JsonSerializerOptions serializerOptions)
+            : this(converterOptions, namingPolicy: null, serializerOptions)
         {
         }
 
-        public EnumConverter(EnumConverterOptions options, JsonNamingPolicy? namingPolicy)
+        public EnumConverter(EnumConverterOptions converterOptions, JsonNamingPolicy? namingPolicy, JsonSerializerOptions serializerOptions)
         {
-            _converterOptions = options;
-            if (namingPolicy != null)
-            {
-                _nameCache = new ConcurrentDictionary<string, string>();
-            }
-            else
+            _converterOptions = converterOptions;
+            _namingPolicy = namingPolicy;
+            _nameCache = new ConcurrentDictionary<ulong, JsonEncodedText>();
+
+            string[] names = Enum.GetNames(TypeToConvert);
+            Array values = Enum.GetValues(TypeToConvert);
+            Debug.Assert(names.Length > 0 && names.Length == values.Length);
```

Comment:
It is possible for enums to contain no values, for instance
```csharp
enum Foo { }

Enum.GetValues(typeof(Foo)).Length // 0
```

```suggestion
            Debug.Assert(names.Length == values.Length);
```

Reply:
Perhaps also add a test for such an enum? It could uncover potential bugs.